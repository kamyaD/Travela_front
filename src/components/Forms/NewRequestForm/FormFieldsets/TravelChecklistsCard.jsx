import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as _ from 'lodash';
import Preloader from '../../../Preloader/Preloader';
import travelChecklistHelper from '../../../../helper/request/RequestUtils';

class TravelChecklistsCard extends Component {
  componentDidMount() {
    const { fetchTravelChecklist, trips } = this.props;
    const tripDestinations = {};
    trips.map(trip => {
      tripDestinations[trip.destination.split(',', 1)] = trip.destination;
    });
    fetchTravelChecklist(null, Object.keys(tripDestinations));
  }
  displayTravelChecklist(checklistItems) {
    return (
      <li className="approval-item" key={checklistItems.id}> 
        <div className="oval" />
        <div className="checklist-name">{checklistItems.name}</div>
      </li>
    );
  }

  render() {
    const { checklistItems, isLoading, userData } = this.props;
    let finalCheckLists = travelChecklistHelper.cleanChecklists(checklistItems, userData);
  
    finalCheckLists = finalCheckLists.reduce((accumulator, checklist) => {
      if(!accumulator.ids.includes(checklist.id)) {
        accumulator.ids.push(checklist.id);
        accumulator.checklist.push(checklist);
      }
      return accumulator;
    }, {ids: [], checklist: []}).checklist;
    return (
      <div>
        <div className="travel-checklist-rectangle">
          <div className="travel-checklist-text">
            <p> Travel Checklist Required For This Trip</p> 
          </div>
          <hr className="travel-checklist-line" />        
          {isLoading
            ? <Preloader /> :  (
              <div className="pending-approvals-block">
                <ul className="approval-list-items">
                  { finalCheckLists && finalCheckLists.map(checklist => {
                    return this.displayTravelChecklist((checklist));
                  })}
                </ul>
              </div>
            )}
        </div>
      </div>
    );
  }
}
TravelChecklistsCard.propTypes = {
  checklistItems: PropTypes.array,
  fetchTravelChecklist: PropTypes.func,
  trips: PropTypes.array,
  isLoading: PropTypes.bool,
  userData: PropTypes.object
};
TravelChecklistsCard.defaultProps = {
  checklistItems: [],
  fetchTravelChecklist: () => {},
  isLoading: false,
  trips: [{}],
  userData: {}
};
export default TravelChecklistsCard;
