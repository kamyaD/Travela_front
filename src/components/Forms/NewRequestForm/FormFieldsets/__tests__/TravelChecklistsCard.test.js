import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import TravelChecklistsCard from '../TravelChecklistsCard';

const props = {
  fetchTravelChecklist: jest.fn(),
  trips: [
    {
      departureDate: '2018-09-20',
      origin: 'Lagos',
      destination: 'Kigali, Rwanda',
    }
  ],
  checklistItems: [
    {
      id: 1,
      name: 'passport',
      checklist: [
        {
          id: 1,
          name: 'passport',
          destinationName: 'Kigali, Rwanda'
        }
      ]
    },
    {
      id: 2,
      name: 'ticketDetails',
      checklist: [{
        id: 2,
        name: 'ticketDetails',
        destinationName: 'Lagos, Nigeria',

      }]
    },
    {
      id: 3,
      name: 'visa',
      checklist: [
        {
          id: 3,
          name: 'visa',
          destinationName: 'Lagos, Nigeria',

        }
      ]
    },
    {
      id: 4,
      name: 'visa',
      destinationName: 'Lagos, Nigeria',

    }
  ],
  userData:{
    id: '2',
    fullName: 'Collins Muru',
    name: 'Collins',
    email: 'collins.muru@andela.com',
    userId: '-LJNw1BsT0LP_E4l2peP',
    passportName: 'Collins',
    department: 'Talent & Development',
    occupation: 'Software Developer',
    manager: 'Collins',
    gender: 'Male',
    createdAt: '2018-09-14T12:48:11.266Z',
    updatedAt: '2018-09-16T07:53:48.835Z',
    roleId: 401938,
    location: 'New'
  },
  trimmedCheckLists: [{
    trip: {
      destinationName: 'Lagos, Nigeria'
    }
  }],
  isLoading: false,
};
describe('<TravelChecklistsCard />',() => {
  const wrapper = mount(
    <MemoryRouter>
      <TravelChecklistsCard {...props} />
    </MemoryRouter>
  );
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should show the loader, while checklist is being fetched', () => {
    const newProps = {
      ...props,
      isLoading: true
    };
    const wrapper = mount(<TravelChecklistsCard {...newProps} />);
    expect(wrapper.find('Preloader')).toHaveLength(1);
  });

  it('should stop showing the loader when checklist is fetched', () =>{
    const wrapper = mount(<TravelChecklistsCard {...props} />);
    expect(wrapper.find('Preloader')).toHaveLength(0);
  });

  it('should render all checkists', () =>{
    const wrapper = mount(<TravelChecklistsCard {...props} />);
    const approvalList = wrapper.find('.approval-list-items');
    expect(approvalList).toHaveLength(1);
    expect(wrapper.find('.checklist-name').at(0).text()).toBe('passport');
    expect(wrapper.find('.checklist-name').at(1).text()).toBe('ticketDetails');
    expect(wrapper.find('.checklist-name').at(2).text()).toBe('visa');
  });
});
