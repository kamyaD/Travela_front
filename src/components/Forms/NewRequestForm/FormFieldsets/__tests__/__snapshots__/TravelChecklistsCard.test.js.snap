// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TravelChecklistsCard /> renders without crashing 1`] = `
<div
  className="mdl-cell mdl-cell--6-col mdl-cell--12-col-tablet mdl-cell--12-phone"
>
  <div
    className="travel-checklist-rectangle"
  >
    <div
      className="travel-checklist-text"
    >
      <p>
         These are the checklist items that you will be required to submit for this trip 
      </p>
    </div>
    <hr
      className="travel-checklist-line"
    />
    <div
      className="pending-approvals-block"
    >
      <div
        className="travel-checklist-loc-title"
      >
        <p>
          Kampala, Uganda
        </p>
      </div>
      <ul
        className="approval-list-items"
      >
        <li
          className="approval-item"
          key="kamp Green card"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            kamp Green card
          </div>
        </li>
        <li
          className="approval-item"
          key="kam Visa"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            kam Visa
          </div>
        </li>
        <li
          className="approval-item"
          key="kam passport"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            kam passport
          </div>
        </li>
        <li
          className="approval-item"
          key="kam Yellow Card"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            kam Yellow Card
          </div>
        </li>
        <li
          className="approval-item"
          key="Travel Ticket Details"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            Travel Ticket Details
          </div>
        </li>
      </ul>
      <div
        className="travel-checklist-loc-title"
      >
        <p>
          Nairobi, Kenya
        </p>
      </div>
      <ul
        className="approval-list-items"
      >
        <li
          className="approval-item"
          key="nair visa"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            nair visa
          </div>
        </li>
        <li
          className="approval-item"
          key="nair passport"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            nair passport
          </div>
        </li>
        <li
          className="approval-item"
          key="nair yellow card"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            nair yellow card
          </div>
        </li>
        <li
          className="approval-item"
          key="nair green card"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            nair green card
          </div>
        </li>
        <li
          className="approval-item"
          key="Travel Ticket Details"
        >
          <div
            className="oval"
          />
          <div
            className="checklist-name"
          >
            Travel Ticket Details
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
